### PR TITLE
Detect multiple enabled payment methods of the same type

### LINKED ISSUE #8384: Implement duplicated payment methods detection mechanism
### Description
Recently, we explored technical approaches to identify duplicate payment methods enabled by plugins on a store. Two solutions were proposed, the simplest one and a more sophisticated one. The simplest solution is based on preparing the hardcoded list of known payment method names used across different plugins and check if the payment method name matches the list multiple times. The more sophisticated approach bases on defining the new contract for plugins to follow by extending the main gateway class. 

Finally, it was concluded that the feature should be developed in two steps
1. The simplest solution will be built in WooPayments. 
2. Once the feature impact is proven, the sophisticated approach should be implemented in WooCommerce. 

This issue is to implement 1️⃣ from the list above and implement it in WooPayments.

### Additional context
* paJDYF-ck0-p2 represents the P2 post, containing both solutions
* The spike PR containing drafts for both solutions is https://github.com/woocommerce/woocommerce/pull/44617
* Main discussion threads are in paJDYF-ck0-p2#comment-23314 and paJDYF-ck0-p2#comment-23652 

### Acceptance criteria
1. Duplicate detection mechanism is developed further in https://github.com/woocommerce/woocommerce/pull/44617, or a separate PR is created, which takes the most relevant pieces from the spike implementation.
2. Duplicates finding implementation from `WC_Gateway_Duplicates_Finder_Static_List` is modified further and reaches production readiness.
3. A notification is implemented that notifies about duplicates found.
    * This step includes confirming the right way to show notification with Product and Design.

If step 3 turns out to significantly impact this feature's complexity, it can be split into a separate issue and worked on after this issue is implemented.

### PR DESCRIPTION
**Pull request by timur27**
Closes https://github.com/Automattic/woocommerce-payments/issues/8384

#### Changes proposed in this Pull Request
This PR is the first iteration of the feature which should help merchants identify payment methods of the same type that are enabled by multiple plugins. It implements the simple solution mentioned in paJDYF-ck0-p2. The inline notice pops up under the payment method and can be dismissed. Once dismissed, it won't appear again upon page refresh. 

> Note: as mentioned in the solution description, the downside of the simple solution implemented in this PR is that we base on the hardcoded keywords list. Due to the fact that there are currently no rules for gateways (internal & third-party) to follow in terms of gateway ID, this solution cannot guarantee its effectiveness in every use case, as some third-party plugin is free to call card gateway in any other way, e.g. `third-party-gateway`. This limitation is known.

For testing steps outlined below, you can use multiple plugins. During development, I used WooPayments, [Mollie](https://wordpress.org/plugins/mollie-payments-for-woocommerce/), [Stripe](https://github.com/woocommerce/woocommerce-gateway-stripe) and [Klarna](https://wordpress.org/plugins/klarna-payments-for-woocommerce/)). If you'd like to quickly setup Mollie, contact @timur27 for test API keys which hopefully will save you a bit of time. For more inspiration on which plugins you could use, please see paJDYF-ck0-p2#comment-23134 which refers the duplicates data.

#### Testing instructions

```
💡 The inline notice text was confirmed with Product & Design and is following: 
This payment method is enabled by other extensions. Review extensions to improve the shopper experience.

Here's how the inline notice looks like:
```
<p align="center">
image
</p>

* Enable credit card by any other plugin e.g. Mollie or Stripe
* Navigate to Payments settings and confirm there's an inline notice under credit card payment method. _Review extensions_ should open the WooCommerce payment settings page
* Enable any other payment method from APMs list. You can use Klarna gateway to check with Klarna or Stripe for any other payment method.
* Confirm that the notice appears.
* Test the scenario when a payment method is enabled in three different places, including WooPayments. At this point, you will see the notice. However, if we disable payment method by WooPayments, the notice should disappear, although it's still enabled by two other plugins. The goal is to show the notice only if WooPayments is part of the plugins group that introduce a duplicate.
* Check Apple Pay/Google Pay by enabling them e.g. under Stripe gateway and checking if the notice appears. Stripe gateway allows PRBs to be enabled with both disabled and enabled CC, and both use-cases should be covered by this PR.
* Dismiss any notice and confirm that it's not shown after closing. Refresh the page and confirm no notice is present.
* Confirm that notice is handled separately, meaning that only part of the notices can be dismissed. After such action, the rest should still be present on the page.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8513` or branch name `add/duplicates-detection` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/duplicates-detection) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f45bac01f9b954400d8d5523c168fe413dc5ea03
- Build time: 2024-04-18 10:38:47 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by timur27** [comment]
@mdmoore I added an optimization in e6a51652a5a5adba241e8198cb7baa21ed8413a0 which allows to share the state of dismissed notices between `ExpressCheckout` and `PaymentMethods` components by using `useContext` (in practice, this means a merchant can dismiss PRB notice & any payment method notice in one go and both will effectively be dismissed). The usage of context also eliminates repeating the code in some places. Just FYI about this addition, I hope it's not interfering with your review much, in case you already started it. Thanks!

File: client/settings/express-checkout/apple-google-pay-item.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@
 import { __ } from '@wordpress/i18n';
 import { Button, CheckboxControl } from '@wordpress/components';
 import interpolateComponents from '@automattic/interpolate-components';
-import React from 'react';
+import React, { useContext, useState } from 'react';
```
Comment (mdmoore) [inline client/settings/express-checkout/apple-google-pay-item.tsx]: Looks like `useState` can be removed here.

**Review by mdmoore — APPROVED** [review]
This all LGTM! Only one minor comment about an unused `useState` hook which GH pointed out.

✅ Navigate to Payments settings and confirm there's an inline notice under credit card payment method. Review extensions should open the WooCommerce payment settings page
    
✅ Enable any other payment method from APMs list. You can use Klarna gateway to check with Klarna or Stripe for any other payment method.
    Confirm that the notice appears.

✅ Test the scenario when a payment method is enabled in three different places, including WooPayments. At this point, you will see the notice. However, if we disable payment method by WooPayments, the notice should disappear, although it's still enabled by two other plugins. The goal is to show the notice only if WooPayments is part of the plugins group that introduce a duplicate.
    
✅ Check Apple Pay/Google Pay by enabling them e.g. under Stripe gateway and checking if the notice appears. Stripe gateway allows PRBs to be enabled with both disabled and enabled CC, and both use-cases should be covered by this PR.

✅ Dismiss any notice and confirm that it's not shown after closing. Refresh the page and confirm no notice is present.
    
✅ Confirm that notice is handled separately, meaning that only part of the notices can be dismissed. After such action, the rest should still be present on the page.

All is testing well and I have no critiques here. 🚢

**Pull request merged**